src/Microsoft.AspNetCore.Blazor.Razor.Extensions/TrimWhitespacePass.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Razor
+{
+    internal class TrimWhitespacePass : IntermediateNodePassBase, IRazorDirectiveClassifierPass
+    {
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (codeDocument == null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+
+            if (documentNode == null)
+            {
+                throw new ArgumentNullException(nameof(documentNode));
+            }
+
+            var method = documentNode.FindPrimaryMethod();
+            if (method != null)
+            {
+                RemoveContiguousWhitespace(method.Children, TraversalDirection.Forwards);
+                RemoveContiguousWhitespace(method.Children, TraversalDirection.Backwards);
+            }
+        }
+
+        private static void RemoveContiguousWhitespace(IntermediateNodeCollection nodes, TraversalDirection direction)
+        {
+            var position = direction == TraversalDirection.Forwards ? 0 : nodes.Count - 1;
+            while (position >= 0 && position < nodes.Count)
+            {
+                var node = nodes[position];
+                bool shouldRemoveNode;
+                bool shouldContinueIteration;
+
+                switch (node)
+                {
+                    case IntermediateToken intermediateToken:
+                        shouldRemoveNode = string.IsNullOrWhiteSpace(intermediateToken.Content);
+                        shouldContinueIteration = shouldRemoveNode;
+                        break;
+
+                    case HtmlContentIntermediateNode htmlContentIntermediateNode:
+                        RemoveContiguousWhitespace(htmlContentIntermediateNode.Children, direction);
+                        shouldRemoveNode = htmlContentIntermediateNode.Children.Count == 0;
+                        shouldContinueIteration = shouldRemoveNode;
+                        break;
+
+                    case CSharpExpressionIntermediateNode _:
+                        shouldRemoveNode = false;
+                        shouldContinueIteration = false; // Because it may produce nonwhitespace output
+                        break;
+
+                    case CSharpCodeIntermediateNode codeIntermediateNode:
```

Comment:
What's the need to remove whitespace inside non-HTML? I'm not sure I understand why that part is important

Reply:
We're not removing whitespace inside non-HTML. This `case` on line 59 is checking whether we hit a `CSharpCodeIntermediateNode` (other than the call to `base.BuildRenderTree`, which we skip). If we do, then we stop the trimming process because we've found something that at runtime might represent non-whitespace.

Reply:
Ok I see that now, I think I misread the code. I was looking at this pretty early in the morning, and I think I got about 4 hours of sleep 😴 

Are you only concerned about *HtmlContent* - or do you care about inside of attributes as well?


You could also consider using `RazorIntermediateNodeWalker` as well if it helps. I think what's here's is OK as it is, but you may need to handle some more cases. Have you looked at what happens when the code is inside a tag helper - inside the body of a component?